tests/NLog.UnitTests/Targets/Wrappers/AsyncTargetWrapperTests.cs

```diff
@@ -108,35 +109,64 @@ private void AsyncTargetWrapperSyncTest_WhenTimeToSleepBetweenBatchesIsEqualToZe
+                    bool blockConsumer = (itemPrepareList.Capacity / 2) == i;  // Force producers to get into blocking-mode
+                    itemPrepareList.Add(logEvent.WithContinuation((ex) => { if (blockConsumer) Thread.Sleep(125); itemWrittenList.Add(sequenceID); }));
                 }
 
-                long startTicks = Environment.TickCount;
-                for (int i = 0; i < itemPrepareList.Count; ++i)
+                AutoResetEvent eventProducer1 = new AutoResetEvent(false);
+                WaitCallback producer = (s) =>
                 {
-                    var logEvent = itemPrepareList[i].Key;
-                    targetWrapper.WriteAsyncLogEvent(logEvent.WithContinuation(itemPrepareList[i].Value));
-                }
+                    AutoResetEvent eventProducer = (AutoResetEvent)s;
+                    if (ReferenceEquals(eventProducer, eventProducer1))
+                        eventProducer1.Set();
+
+                    int indexStart = ReferenceEquals(eventProducer, eventProducer1) ? 1 : 0;
+                    for (int i = indexStart; i < itemPrepareList.Count; i += 2)
+                    {
+                        targetWrapper.WriteAsyncLogEvent(itemPrepareList[i]);
+                    }
+
+                    if (ReferenceEquals(eventProducer, eventProducer1))
+                        eventProducer1.Set();
+                };
+
+                ThreadPool.QueueUserWorkItem(producer, eventProducer1);
+                Assert.True(eventProducer1.WaitOne(5000));
+
+                long startTicks = Environment.TickCount;
+
+                producer(null);
+                Assert.True(eventProducer1.WaitOne(5000));
+
+                long elapsedMilliseconds = Environment.TickCount - startTicks;
 
                 targetWrapper.Flush(flushHandler);
 
                 for (int i = 0; i < itemPrepareList.Count * 2 && itemWrittenList.Count != itemPrepareList.Count; ++i)
                     Thread.Sleep(1);
 
-                long elapsedMilliseconds = Environment.TickCount - startTicks;
-
                 Assert.Equal(itemPrepareList.Count, itemWrittenList.Count);
-                int prevSequenceID = 0;
-                for (int i = 0; i < itemWrittenList.Count; ++i)
+
+                int writer1sequenceID = itemWrittenList[0];
+                int writer2sequenceID = 0;
+                for (int i = 1; i < itemWrittenList.Count; ++i)
                 {
-                    Assert.True(prevSequenceID < itemWrittenList[i]);
-                    prevSequenceID = itemWrittenList[i];
+                    if (writer1sequenceID < itemWrittenList[i])
+                    {
+                        Assert.True(writer1sequenceID < itemWrittenList[i]);
+                        writer1sequenceID = itemWrittenList[i];
+                    }
+                    else
+                    {
+                        Assert.True(writer2sequenceID < itemWrittenList[i]);
```

Comment:
Please add a message, as we have multiple assert.trues in the test method (and if they fail, it will be unclear which one)

Reply:
Made the code more readable.